### ISSUE TITLE
Run sonobuoy cli with custom plugins from local filesystem

### ISSUE THREAD

**Issue by mwieczorek** [heptio/sonobuoy#405]
**Is this a BUG REPORT or FEATURE REQUEST?**:
Feature request

**What happened**:
I'd like to `sonobuoy run` with custom plugins. Right now it's not possible. I need to:
1. run `sonobuoy gen` > file.yaml
2. modify configmap with plugins definitions
3. kubectl apply -f file.yaml

**What you expected to happen**:
I'd like to us only sonobuoy cli without editing yaml files.
F.e.:
1. I have custom config.json with list of the plugins
2. I have folder with my plugin definitions, each in a separate yaml file
3. I run:
`sonobuoy run --config config.json --pluginsFolder <path_to_folder>`
4. sonobuoy would get files from folder, parse and validate it, and modify configmap 'in fly' before hitting apiserver

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Sonobuoy tarball (which contains * below)
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):

**Comment by hoegaarden** [comment]
Currently I am [post-processing][x] the output of `sonobuoy gen ...` with some `jq`/`yq` magic. Doing that sort of thing in the `sonobuoy` binary would be a great UX improvement.

[x]: https://github.com/pivotal-k8s/chkrootkit-sonobuoy-plugin/blob/7649b09ce699b5ace9eda00ed42931441956d663/sonobuoy-gen.sh#L13-L32

**Comment by johnSchnake** [comment]
One thing that seems unusual/unnecessary/unnatural is the way the plugins are loaded/created by the master/aggregator.

For one, it seems unnecessary to me why we are allowing for the customization of the plugin search directory. Since we control the whole flow and where the mounts go, it seems like we could just have removed those options.

One of the issues I see about extending the UI here is how to deal with the "duplication" of plugin info in the yaml. The `config.json` has a list of all the plugins we've "selected" and then we also create config maps for a set of plugins. Right now those are completely decoupled and the master/aggregator does this:
 - loads the config.json and sees selected plugins and plugin search path
 - loads all plugins in the search path
 - filters the loaded plugins to only have the selected ones

It seems to me the reason why those are decoupled is because the filename is not necessarily indicative of the plugin name (which is a field in the yaml).

I'd rather just improve the logic around the plugins configmaps that are mounted into the container and then anything that is loaded is run (I dont see why we have to have a duplicate list of "selected" plugins if we just ensure to only make configmaps for the selected ones in the first place)

Once we settle that I think we can add flags pretty easily for just pointing to a list of files you want to load as plugins.

Then a final thing to consider is how to solve this use case: you want to run run a custom list of plugins and include just e2e and your plugin. Using the flag described above to point to your plugin file would add yours, but then you would have to decide how to add the e2e plugin. Locally you wont have the yaml for that so we seem to have a few options:
 - add separate flags for plugins which exist internal to sonobuoy (e2e and systemd_logs) (e.g. `sonobuoy run --plugins e2e --plugins-file mypluginDef.yaml`
 - add separate flags for each of the internal plugins which would be a boolean (`sonobuoy run --e2e --systemd_logs=false --plugins-file mypluginDef.yaml`
 - add special keywords which the new plugins flag will recognize and treat differently (`sonobuoy run --plugins-file mypluginDef.yaml --plugins-file e2e`)

I'm going to put a design doc together for this because there are details to work out but I thought I would just start with a description here of some of the ideas/issues

**Comment by stevesloka** [comment]
I think dropping the config.json with respect to plugins would be a good move. I like the idea of Sonobuoy searching for plugins in a location(s), this way we just need to drop in configuration plugin files and sonobuoy will just pick them up.

**Comment by johnSchnake** [comment]
Wrote design doc here, please take a look and comment. I think this should be pretty straight forward if ya'll agree with the plan. I've been looking at the code a lot today on this and dont anticipate big problems/complications.
https://docs.google.com/document/d/1TN_k0TlOmoqHv4QCJWKVt_WmSq9o5LWPRlvcIFLA4R4/edit?usp=sharing

tl;dr; I wrote this up to go with the 3rd option I listed. A single flag with 2 special cases. I think for the user that is pretty straightforward rather than having multiple flags.
